File: 1-100/21-merge-two-sorted-lists.cpp
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};


ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
    ListNode *dummyHead = new ListNode(0);
    ListNode *curr = dummyHead;

    while (list1 && list2)
    {
        if (list1->val < list2->val){
            curr->next = new ListNode(list1->val);
            list1 = list1->next;
        }
        else{
            curr->next = new ListNode(list2->val);
            list2 = list2->next;
        }
        curr = curr->next;
    }
    if (list1){
        curr->next = list1;
    }
    else if (list2){
        curr->next = list2;
    }

    return dummyHead->next;
}

File: python version/21-merge-two-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        dummy_head = ListNode()
        curr = dummy_head
        l1 = list1
        l2 = list2
        while l1 and l2:
            if l1.val < l2.val:
                new_val = l1.val
                l1 = l1.next
            else:
                new_val = l2.val
                l2 = l2.next
            curr.next = ListNode(new_val)
            curr = curr.next
        if l1:
            curr.next = l1
        if l2:
            curr.next = l2 
        return dummy_head.next
            